Implemented check_disk_space function

martinbrose opened the pull request:
Addressing #1551.

- #1551

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_1590). All of your documentation changes will be reflected on that endpoint.

Wauplin commented on src/huggingface_hub/file_download.py:
```diff
@@ -961,6 +961,49 @@ def repo_folder_name(*, repo_id: str, repo_type: str) -> str:
+    if not has_enough_space:
+        if target_dir is not None:
+            logger.warning(
+                "Not enough free disk space to download the file. The expected file size is: {:,.2f} MB. Temporary"
+                " file path: {} ({:,.2f} MB free), target directory: {} ({:,.2f} MB free)".format(
+                    expected_size / 1e6, temp_dir, temp_dir_size / 1e6, target_dir, target_dir_size / 1e6
+                )
```
`"...".format(...)` is valid but for readability and consistency within `huggingface_hub`, I'd rather use f-strings `f"..."` 

([here](https://stackoverflow.com/a/43123579) is a SO post about it, TL;DR: f-string are usually faster but that's not why we use them. The decision has to be taken based on readability and consistency across a codebase).

Wauplin commented on src/huggingface_hub/file_download.py:
```diff
@@ -1394,6 +1437,13 @@ def _resumable_file_manager() -> Generator[io.BufferedWriter, None, None]:
         with temp_file_manager() as temp_file:
             logger.info("downloading %s to %s", url, temp_file.name)
 
+            if local_dir is None:
+                target_dir = os.path.dirname(blob_path)
+            else:
+                target_dir = os.path.dirname(local_dir)
+
+            _check_disk_space(expected_size, temp_file, target_dir)
```
With the update that I suggested above, the logic here change slightly to something like this:

```py
            if expected_size is not None: # might be None if HTTP header not set correctly
                # Check tmp path
                _check_disk_space(expected_size, os.path.dirname(temp_file.name))

                # Check destination
                _check_disk_space(expected_size, os.path.dirname(blob_path))
                if local_dir is not None:
                    _check_disk_space(expected_size, local_dir)
```

Wauplin reviewed (changes requested):
Hey @martinbrose thanks again for another PR! Your approach here is good. I focused the review and comments on how the logic could be simplified by having a helper doing only 1 check but being called multiple times. This simplifies the overall logic by reducing the number of `if ... is not None:` statements. Let me know what you think about it :)

martinbrose commented:
@Wauplin many thanks for your detailed review comments. I'll attend to them soon.

What I realised is that I haven't added any test cases to `test_file_download.py`...
Would you have pointers for creating test file storage of a specific size in the testing environment?
So I could check for cases where enough space is available and where there is not?

Would you also want to have these helpers added to `__init__.py`?

Wauplin commented:
Good questions!

> Would you have pointers for creating test file storage of a specific size in the testing environment?
So I could check for cases where enough space is available and where there is not?

What I would do is to test the helper individually and not the end-to-end process of downloading a file.

1. I would define a new class `TestDiskUsageWarning(unittest.TestCase)` in `test_file_download.py`
2. I would 1 test where the free disk space is ok, 1 test where the free disk space is too low
3. Instead of artificially reducing the disk space to make the tests work, I would mock the `shutil.disk_usage` method.

To mock functions, you can use [`unittest.mock.patch`](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.patch). Instead of running `disk_usage`, the test will simply use the mock value. This is done for example [here](https://github.com/huggingface/huggingface_hub/blob/b9901bf71839850c10ea42162c69aecabc96a46b/tests/test_file_download.py#L901). In your case, you will decorate your test with `@patch("huggingface_hub.file_download.shutil.disk_usage")`. Then the test takes as input a `Mock` object on which you can set a specific return value (for instance: `mock.return_value.free = 1024 * 1024` to simulate the fact that there is 1MB left on disk). When you call `_check_disk_space` afterwards, it will use this mocked value. Hope these explanations are clear enough. You can also look at the [Python docs](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.patch) and check some examples in `huggingface_hub`.

> Would you also want to have these helpers added to __init__.py?

No need for that! It's a small private helper so it's better to not expose it :)

martinbrose commented:
Hi @Wauplin,

I believe I have addressed all your review points.

The only remaining point from my side:
When I run `make quality`, I receive the following messages:

```
src/huggingface_hub/file_download.py:977: error: Unsupported operand types for >= ("int" and "None")  [operator]
src/huggingface_hub/file_download.py:977: note: Right operand is of type "int | None"
src/huggingface_hub/file_download.py:981: error: Unsupported operand types for / ("None" and "float")  [operator]
src/huggingface_hub/file_download.py:981: note: Left operand is of type "int | None"
src/huggingface_hub/file_download.py:986: error: No return value expected  [return-value]
```

Is this something I need to address?

Thank you!

martinbrose commented:
Oh, actually another question:
I did add a third test case when free disk space is exactly the size of the expected file size.
Do you think that adds value or not?

Wauplin commented on src/huggingface_hub/file_download.py:
```diff
@@ -961,6 +961,31 @@ def repo_folder_name(*, repo_id: str, repo_type: str) -> str:
     return REPO_ID_SEPARATOR.join(parts)
 
 
+def _check_disk_space(expected_size: Optional[int], target_dir: Union[str, Path]) -> None:
```
```suggestion
def _check_disk_space(expected_size: int, target_dir: Union[str, Path]) -> None:
```

Your problem comes from the fact that you expect `int` or `None` as input value. Mypy tells you that if a None value is passed, the line `target_dir_free >= expected_size` will fail. This is how static typing is useful: mypy doesn't even need to run the code to tell you there is a bug. In this case, just change the input type and that's it! :)

Wauplin commented on src/huggingface_hub/file_download.py:
```diff
@@ -961,6 +961,31 @@ def repo_folder_name(*, repo_id: str, repo_type: str) -> str:
+    """Check disk usage and log a warning if there is not enough disk space to download the file.
+
+    Args:
+        expected_size (`int`, *optional*):
```
```suggestion
        expected_size (`int`):
```

Wauplin commented on src/huggingface_hub/file_download.py:
```diff
@@ -961,6 +961,31 @@ def repo_folder_name(*, repo_id: str, repo_type: str) -> str:
+            f"The expected file size is: {expected_size / 1e6:.2f} MB. "
+            f"The target location {target_dir} only has {target_dir_free / 1e6:.2f} MB free space."
+        )
+
+    return has_enough_space
```
```suggestion
```

No need for a return value. I believe you added this return value to make your tests pass. A better way to do this is to use [`self.assertWarns`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertWarns) when you expect a warning to be raised. There are other occurrences of this in the tests if you want to have a look.

Wauplin commented on tests/test_file_download.py:
```diff
@@ -84,6 +85,30 @@
+    def setUpClass(cls):
+        # Test with 100MB expected file size
+        cls.expected_size = 100 * 1024 * 1024
+        cls.disk_usage_mock = Mock()
+
+    @classmethod
+    def tearDownClass(cls) -> None:
+        cls.disk_usage_mock.reset_mock()
```
```suggestion
```

No need for a class-level mock. The `disk_usage_mock` value is already injected for you by the `@patch` decorator.

Wauplin reviewed (commented):
Thanks for the changes @martinbrose !
You mypy issues can be solved quite easily by changing the expected input type.
About the test case when you have exactly the good amount of free space, I don't think it's worth it. TBH on a normal computer there are 0 chances that you'll use exactly the 100% of your hard-drive. And even if that's the case your computer will have other issues since it's always writing/reading the disk. So let's not overthink this :)

martinbrose commented:
Addressed nearly all of the review points.

I need to read up on testing for warnings with `self.assertWarns` and address the outstanding right afterward.

martinbrose commented:
I suspect we don't really have a warning with `logger.warning`? I might need to provide a mock logger and check that one as part of the test.

Wauplin commented on src/huggingface_hub/file_download.py:
```diff
@@ -961,6 +961,31 @@ def repo_folder_name(*, repo_id: str, repo_type: str) -> str:
+    has_enough_space = target_dir_free >= expected_size
+
+    if not has_enough_space:
+        logger.warning(
```
Ah yes! You should use `warnings.warn` here for a proper warning message. This is different than logger.warning logs. Glad that you've implemented tests to realize that!

martinbrose replied:
Ok, if `warnings.warn` is okay with you, it becomes much easier.

However, I'm at a loss on how to test for no warning with `self.assertWarns`.

I propose the following (with `import warnings` earlier in the file):
```python
class TestDiskUsageWarning(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        # Test with 100MB expected file size
        cls.expected_size = 100 * 1024 * 1024

    @patch("huggingface_hub.file_download.shutil.disk_usage")
    def test_disk_usage_warning(self, disk_usage_mock: Mock) -> None:
        # Test with only 1MB free disk space / not enough disk space,  with UserWarning expected
        disk_usage_mock.return_value.free = 1024 * 1024
        with warnings.catch_warnings(record=True) as w:
            # Cause all warnings to always be triggered.
            warnings.simplefilter("always")
            _check_disk_space(expected_size=self.expected_size, target_dir=disk_usage_mock)
            assert len(w) == 1
            assert issubclass(w[-1].category, UserWarning)

        # Test with 200MB free disk space / enough disk space,  with no warning expected
        disk_usage_mock.return_value.free = 200 * 1024 * 1024
        with warnings.catch_warnings(record=True) as w:
            # Cause all warnings to always be triggered.
            warnings.simplefilter("always")
            _check_disk_space(expected_size=self.expected_size, target_dir=disk_usage_mock)
            assert len(w) == 0
```

Wauplin replied:
Really good way to test the absence of warnings! So yes, go ahead with this solution since it works :)

Wauplin reviewed (approved):
Great work again! Looks good to merge :)
